### DISCUSSION TITLE
Diffusion of Constant Monomials

### DISCUSSION THREAD

**Discussion by lloyd-5731** [idaholab/moose#23143] [Q&A General] [answered]
Is it possible to diffuse a constant monomial variable?
I have an aux variable that is derived from a material value, and hence it must be a constant monomial. (is this true?)
I am looking to parse this value into a sub app to diffuse it, but it appears that my auxvar that is a constant monomial will not diffuse using `Diffusion` and `TimeDerivative` kernels.
When I change the variable to a Lagrange (which will not work with the parent app) it diffuse fine.

Is there a work around for this?

**Comment by GiudGiud** [comment]
Hello

Yes it is possible. However it must be a nonlinear or primary variable. It cannot be an auxiliary variable
Can you please attach your input here ?

Guillaume

**Comment by lloyd-5731** [comment]
Here is my parent input file:

```
P0 = 1      # Injection pressure
rho = 1       # Density
k = 1       # Permeability
phi = 0.5     # Porosity
T = 298
mu = 0.01
init_variable = 0  
input_from_BC = 1


[Mesh]
    [plate]
        type = GeneratedMeshGenerator
        dim = 2
        xmax = 0.1
        nx = 100
        ymax = 0.1
        ny = 100
    []
[]
[UserObjects]
    [dictator]
        type = PorousFlowDictator
        porous_flow_vars = 'porepressure'
        number_fluid_phases = 1
        number_fluid_components = 1
    []
    [pc]
        type = CustomCapillaryPressure
        p_max = 1000
        epsilon = 80
        p_min = 0
        log_extension = false
    []
[]

[GlobalParams]
    PorousFlowDictator = dictator
    gravity = '0 0 0'
[]

[Variables]
    [porepressure]
        initial_condition = 0
    []
    [variable_to_advect]
        initial_condition = ${init_variable}
    []
[]

[Kernels]
    [dmdt]
        type = PorousFlowMassTimeDerivative
        variable = porepressure
    []
    [flux]
        type = PorousFlowAdvectiveFlux
        variable = porepressure
    []
    [advect_var_flux]
        type = ConservativeAdvection
        variable = variable_to_advect
        velocity = sat_vel
        upwinding_type = full
    []
    [advect_var_time_derivative]
        type = ADTimeDerivative
        variable = variable_to_advect
    []
[]

[AuxVariables]
    [saturation]
        family = MONOMIAL
        order = CONSTANT
    []
    [vel_x]
        order = CONSTANT
        family = MONOMIAL
    []
[]

[AuxKernels]
    [saturation]
        type = PorousFlowPropertyAux
        variable = saturation
        property = saturation
    []
    [vel_x]
        type = PorousFlowDarcyVelocityComponent
        variable = vel_x
        component = x
        porepressure = porepressure
    []
   []

[BCs]
    [injection]
        type = DirichletBC
        variable = porepressure
        boundary = 'top'
        value = ${P0}
    []

    [Var_in]
        type = DirichletBC
        variable = variable_to_advect
        boundary = 'top'
        value = ${input_from_BC}
    []
[]

[Modules]
    [FluidProperties]
        [resin]
            type = SimpleFluidProperties
            density0 = ${rho}
            viscosity = ${mu}
        []
    []
[]

[Materials]
    [porosity]
        type = PorousFlowPorosity
        porosity_zero = ${phi}
    []
    [permeability]
        type = PorousFlowPermeabilityConst
        permeability = '${k} 0 0   0 ${k} 0   0 0 ${k}'
    []
    [saturation_calculator]
        type = PorousFlow1PhaseP
        porepressure = porepressure
        capillary_pressure = pc
    []
    [temperature]
        type = PorousFlowTemperature
        temperature = ${T}
    []
    [massfrac]
        type = PorousFlowMassFraction
    []
    [resin]
        type = PorousFlowSingleComponentFluid
        fp = resin
        phase = 0
    []
    [relperm]
        type = PorousFlowRelativePermeabilityCorey
        n = 3
        s_res = 0.1
        sum_s_res = 0.1
        phase = 0
    []
    [darcy_velocity]
        type = PorousFlowDarcyVelocityMaterial
    []

[]

[Preconditioning]
    [smp]
        type = SMP
        full = true
        petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
        petsc_options_value = ' lu       mumps'
    []
    
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    end_time = 0.5
    [TimeStepper]
        type = SolutionTimeAdaptiveDT
        dt = 0.00001
    []

    automatic_scaling = true
    nl_abs_tol = 1e-10
    nl_rel_tol = 1e-10
[]

[MultiApps]
    [sub]
      type = FullSolveMultiApp
      input_files = sub.i
      execute_on = TIMESTEP_END
    []
[]
  
 [Transfers]
    [to_sub_x]
      type = MultiAppCopyTransfer
      source_variable = vel_x
      variable = u
      multi_app = sub
      direction = to_multiapp
      execute_on = TIMESTEP_END
    []
    [from_sub_x]
        type = MultiAppCopyTransfer
        source_variable = u
        variable = vel_x
        multi_app = sub
        direction = from_multiapp
        execute_on = TIMESTEP_END
    []
[]
  
[Outputs]
    exodus = true
    print_linear_residuals = false
[]
```

**Comment by lloyd-5731** [comment]
And my child file:
```
[Mesh]
    [plate]
        type = GeneratedMeshGenerator
        dim = 2
        xmax = 0.1
        nx = 100
        ymax = 0.1
        ny = 100
    []
[]
[Variables]
    [u]
        order = CONSTANT
        family = MONOMIAL
    []
    
[]

[Kernels]
    [diff_u]
        type = Diffusion
        variable = u
    []
    [deriv_u]
        type = TimeDerivative
        variable = u
    []
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    end_time = 1e-4
    dt =  1E-7
  
    automatic_scaling = true
    nl_abs_tol = 1e-10
    nl_rel_tol = 1e-10
[]
[Outputs]
    exodus = true
    print_linear_residuals = false
[]
```

**Comment by lloyd-5731** [comment]
I need to take the value of the `DarcyVelocity` which comes out as an `AuxVar` (Using the existing `PorousFlowDarcyVelocityComponent` Kernel), slightly diffuse this value (which I have tried to do in my SubApp), and then feed that back in to use later in the calculation.

Any thoughts would be appreciated. Please note this is a MWE.

**Comment by hugary1995** [comment, marked as answer]
You can project the Monomial from the main app to a Lagrange variable in the subapp. @GiudGiud may know the best transfer to use for that. I'd start with trying maybe MultiAppProjectionTransfer or MultiAppMeshFunctionTransfer.

**Reply by lloyd-5731** [reply]
It look like `MultiAppProjectionTransfer` was the solution.
That's now looking much more like what I want it to thank you.

My only other question is that I want to process the variable vel_x in my SubApp (which is now working)
But I want to feed that back in the calculation so that the value can be combined with sat_vel, that can be used in the advection.

I think at the moment I am just performing this diffusion at the end of the time step so it will not be able to be brought into sat_vel.

Any ideas?

**Reply by GiudGiud** [reply]
If you want to diffuse in the same time step and use that diffused result in that same time step you have two options: 

you include this variable and it’s equation system in the patent application solve and get rid of the multi app.  This is called full coupling, solving all variables in the same non linear solve
OR
you use fixed point iterations between the parent app and the child app to achieve tight coupling. The main app will rerun its time step with the updated velocity. To do that you will need to specify fixed point iteration parameters in the Executioner block. 

Guillaume

**Reply by lloyd-5731** [reply]
Thanks Guilaume, it sounds like I need to try the fixed point iteration in that case. 

Is there an example file on implementing `fixed point iterations`? As I can't seem to find one in the docs apart from this page: https://mooseframework.inl.gov/bison/syntax/Executioner/FixedPointAlgorithms/index.html

Lloyd

**Reply by GiudGiud** [reply]
There s examples in moose/test/tests/multiapps

**Reply by lloyd-5731** [reply]
Ok thank you, for the help Guilaume.
When I try and implement it gives me this error:
```*** ERROR ***
The following error occurred in the object "sub", of type "FullSolveMultiApp".

FullSolveMultiApp is not compatible with auto_advance=false
```

I am knew to using a multi app system so thank you for being patient!

**Reply by GiudGiud** [reply]
A FullSolveMultiAPp runs an entire simulation, not a single timestep. 
If you want that, then set auto_advance = true

If not then you need to switch to a TransientMultiAPp

**Reply by lloyd-5731** [reply]
Sorted. Thank you.
